File: plugins/flutter_aepoptimize/ios/Classes/FlutterAEPOptimizePlugin.m
/*
Copyright 2025 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

@import AEPOptimize;
@import AEPCore;
@import Foundation;
#import "FlutterAEPOptimizePlugin.h"
#import "FlutterAEPOptimizeDataBridge.h"

@interface FlutterAEPOptimizePlugin ()
@property(nonatomic, strong) FlutterMethodChannel *channel;
@end

@implementation FlutterAEPOptimizePlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
    FlutterMethodChannel *channel =
        [FlutterMethodChannel methodChannelWithName:@"flutter_aepoptimize"
                                    binaryMessenger:[registrar messenger]];
    FlutterAEPOptimizePlugin *instance = [[FlutterAEPOptimizePlugin alloc] init];
    instance.channel = channel;
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    if ([@"extensionVersion" isEqualToString:call.method]) {
        result([AEPMobileOptimize extensionVersion]);
    } else if ([@"updatePropositions" isEqualToString:call.method]) {
        [self handleUpdatePropositions:call result:result];
    } else if ([@"getPropositions" isEqualToString:call.method]) {
        [self handleGetPropositions:call result:result];
    } else if ([@"registerOnPropositionsUpdate" isEqualToString:call.method]) {
        [self handleRegisterOnPropositionsUpdate:result];
    } else if ([@"clearCachedPropositions" isEqualToString:call.method]) {
        [AEPMobileOptimize clearCachedPropositions];
        result(nil);
    } else if ([@"offerDisplayed" isEqualToString:call.method]) {
        [self handleOfferDisplayed:call result:result];
    } else if ([@"offerTapped" isEqualToString:call.method]) {
        [self handleOfferTapped:call result:result];
    } else if ([@"generateDisplayInteractionXdm" isEqualToString:call.method]) {
        [self handleGenerateDisplayInteractionXdm:call result:result];
    } else if ([@"generateTapInteractionXdm" isEqualToString:call.method]) {
        [self handleGenerateTapInteractionXdm:call result:result];
    } else if ([@"generateReferenceXdm" isEqualToString:call.method]) {
        [self handleGenerateReferenceXdm:call result:result];
    } else if ([@"batchDisplayed" isEqualToString:call.method]) {
        [self handleBatchDisplayed:call result:result];
    } else if ([@"batchGenerateDisplayInteractionXdm" isEqualToString:call.method]) {
        [self handleBatchGenerateDisplayInteractionXdm:call result:result];
    } else {
        result(FlutterMethodNotImplemented);
    }
}

#pragma mark - API Handlers

- (void)handleUpdatePropositions:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSDictionary *arguments = call.arguments;
    NSArray<AEPDecisionScope *> *scopes = [FlutterAEPOptimizeDataBridge decisionScopesFromArray:arguments[@"decisionScopes"]];

    if (!scopes || scopes.count == 0) {
        result([FlutterError errorWithCode:@"INVALID_ARGUMENT"
                                   message:@"decisionScopes is required"
                                   details:nil]);
        return;
    }

    NSDictionary *xdm = [arguments[@"xdm"] isKindOfClass:[NSDictionary class]] ? arguments[@"xdm"] : nil;
    NSDictionary *data = [arguments[@"data"] isKindOfClass:[NSDictionary class]] ? arguments[@"data"] : nil;
    NSNumber *timeoutNumber = [arguments[@"timeout"] isKindOfClass:[NSNumber class]] ? arguments[@"timeout"] : nil;

    void (^completionHandler)(NSDictionary<AEPDecisionScope *, AEPOptimizeProposition *> * _Nullable, NSError * _Nullable) =
        ^(NSDictionary<AEPDecisionScope *, AEPOptimizeProposition *> * _Nullable propositions, NSError * _Nullable error) {
            if (error) {
                result([self flutterErrorFromNSError:error]);
            } else {
                result([FlutterAEPOptimizeDataBridge dictionaryFromPropositionsMap:propositions]);
            }
        };

    if (timeoutNumber && ![timeoutNumber isKindOfClass:[NSNull class]]) {
        NSTimeInterval timeout = [timeoutNumber doubleValue];
        // Note: the SDK's generated ObjC header has swapped parameter names for timeout/andData.
        // The `timeout:` selector slot takes the data dictionary, and `andData:` takes the NSTimeInterval.
        [AEPMobileOptimize updatePropositions:scopes
                                      withXdm:xdm
                                      timeout:data
                                      andData:timeout
                                   completion:completionHandler];
    } else {
        [AEPMobileOptimize updatePropositions:scopes
                                      withXdm:xdm
                                      andData:data
                                   completion:completionHandler];
    }
}

- (void)handleGetPropositions:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSDictionary *arguments = call.arguments;
    NSArray<AEPDecisionScope *> *scopes = [FlutterAEPOptimizeDataBridge decisionScopesFromArray:arguments[@"decisionScopes"]];

    if (!scopes || scopes.count == 0) {
        result([FlutterError errorWithCode:@"INVALID_ARGUMENT"
                                   message:@"decisionScopes is required"
                                   details:nil]);
        return;
    }

    NSNumber *timeoutNumber = [arguments[@"timeout"] isKindOfClass:[NSNumber class]] ? arguments[@"timeout"] : nil;

    void (^completionHandler)(NSDictionary<AEPDecisionScope *, AEPOptimizeProposition *> * _Nullable, NSError * _Nullable) =
        ^(NSDictionary<AEPDecisionScope *, AEPOptimizeProposition *> * _Nullable propositions, NSError * _Nullable error) {
            if (error) {
                result([self flutterErrorFromNSError:error]);
            } else {
                result([FlutterAEPOptimizeDataBridge dictionaryFromPropositionsMap:propositions]);
            }
        };

    if (timeoutNumber) {
        [AEPMobileOptimize getPropositions:scopes timeout:[timeoutNumber doubleValue] completion:completionHandler];
    } else {
        [AEPMobileOptimize getPropositions:scopes completion:completionHandler];
    }
}

- (void)handleRegisterOnPropositionsUpdate:(FlutterResult)result {
    [AEPMobileOptimize onPropositionsUpdate:^(NSDictionary<AEPDecisionScope *, AEPOptimizeProposition *> * _Nonnull propositions) {
        NSDictionary *encoded = [FlutterAEPOptimizeDataBridge dictionaryFromPropositionsMap:propositions];
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.channel invokeMethod:@"onPropositionsUpdate" arguments:encoded];
        });
    }];
    result(nil);
}

- (void)handleOfferDisplayed:(FlutterMethodCall *)call result:(FlutterResult)result {
    AEPOptimizeProposition *proposition = [FlutterAEPOptimizeDataBridge propositionFromOfferTrackingDictionary:call.arguments];
    AEPOffer *offer = proposition.offers.firstObject;
    if (offer) {
        [offer displayed];
    }
    result(nil);
}

- (void)handleOfferTapped:(FlutterMethodCall *)call result:(FlutterResult)result {
    AEPOptimizeProposition *proposition = [FlutterAEPOptimizeDataBridge propositionFromOfferTrackingDictionary:call.arguments];
    AEPOffer *offer = proposition.offers.firstObject;
    if (offer) {
        [offer tapped];
    }
    result(nil);
}

- (void)handleGenerateDisplayInteractionXdm:(FlutterMethodCall *)call result:(FlutterResult)result {
    AEPOptimizeProposition *proposition = [FlutterAEPOptimizeDataBridge propositionFromOfferTrackingDictionary:call.arguments];
    AEPOffer *offer = proposition.offers.firstObject;
    if (offer) {
        result([offer generateDisplayInteractionXdm]);
    } else {
        result(nil);
    }
}

- (void)handleGenerateTapInteractionXdm:(FlutterMethodCall *)call result:(FlutterResult)result {
    AEPOptimizeProposition *proposition = [FlutterAEPOptimizeDataBridge propositionFromOfferTrackingDictionary:call.arguments];
    AEPOffer *offer = proposition.offers.firstObject;
    if (offer) {
        result([offer generateTapInteractionXdm]);
    } else {
        result(nil);
    }
}

- (void)handleGenerateReferenceXdm:(FlutterMethodCall *)call result:(FlutterResult)result {
    AEPOptimizeProposition *proposition = [FlutterAEPOptimizeDataBridge propositionFromDictionary:call.arguments];
    if (proposition) {
        result([proposition generateReferenceXdm]);
    } else {
        result(nil);
    }
}

- (void)handleBatchDisplayed:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSMutableArray<AEPOptimizeProposition *> *propositions = [NSMutableArray array];
    NSMutableArray<AEPOffer *> *offers = [NSMutableArray array];
    for (NSDictionary *dict in call.arguments) {
        AEPOptimizeProposition *prop = [FlutterAEPOptimizeDataBridge propositionFromOfferTrackingDictionary:dict];
        if (prop && prop.offers.count > 0) {
            [propositions addObject:prop];
            [offers addObject:prop.offers[0]];
        }
    }
    if (offers.count > 0) {
        [AEPMobileOptimize displayed:offers];
    }
    result(nil);
}

- (void)handleBatchGenerateDisplayInteractionXdm:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSMutableArray<AEPOptimizeProposition *> *propositions = [NSMutableArray array];
    NSMutableArray<AEPOffer *> *offers = [NSMutableArray array];
    for (NSDictionary *dict in call.arguments) {
        AEPOptimizeProposition *prop = [FlutterAEPOptimizeDataBridge propositionFromOfferTrackingDictionary:dict];
        if (prop && prop.offers.count > 0) {
            [propositions addObject:prop];
            [offers addObject:prop.offers[0]];
        }
    }
    if (offers.count > 0) {
        result([AEPMobileOptimize generateDisplayInteractionXdm:offers]);
    } else {
        result(nil);
    }
}

#pragma mark - Helpers

- (FlutterError *)flutterErrorFromNSError:(NSError *)error {
    return [FlutterError errorWithCode:[NSString stringWithFormat:@"%ld", (long)error.code]
                               message:error.localizedDescription
                               details:error.domain];
}

@end

File: plugins/flutter_aepoptimize/android/src/main/java/com/adobe/marketing/mobile/flutter/flutter_aepoptimize/FlutterAEPOptimizePlugin.java
/*
Copyright 2025 Adobe. All rights reserved.
This file is licensed to you under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License. You may obtain a copy
of the License at http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
OF ANY KIND, either express or implied. See the License for the specific language
governing permissions and limitations under the License.
*/

package com.adobe.marketing.mobile.flutter.flutter_aepoptimize;

import com.adobe.marketing.mobile.AdobeCallbackWithError;
import com.adobe.marketing.mobile.AdobeError;
import com.adobe.marketing.mobile.optimize.DecisionScope;
import com.adobe.marketing.mobile.optimize.Offer;
import com.adobe.marketing.mobile.optimize.OfferUtils;
import com.adobe.marketing.mobile.optimize.Optimize;
import com.adobe.marketing.mobile.optimize.OptimizeProposition;

import androidx.annotation.NonNull;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class FlutterAEPOptimizePlugin implements FlutterPlugin, MethodCallHandler {

    private MethodChannel channel;

    @Override
    public void onAttachedToEngine(@NonNull final FlutterPluginBinding binding) {
        channel = new MethodChannel(binding.getBinaryMessenger(), "flutter_aepoptimize");
        channel.setMethodCallHandler(this);
    }

    @Override
    public void onDetachedFromEngine(@NonNull final FlutterPluginBinding binding) {
        if (channel != null) {
            channel.setMethodCallHandler(null);
        }
    }

    @Override
    public void onMethodCall(MethodCall call, @NonNull Result result) {
        if ("extensionVersion".equals(call.method)) {
            result.success(Optimize.extensionVersion());
        } else if ("updatePropositions".equals(call.method)) {
            handleUpdatePropositions(call, result);
        } else if ("getPropositions".equals(call.method)) {
            handleGetPropositions(call, result);
        } else if ("registerOnPropositionsUpdate".equals(call.method)) {
            handleRegisterOnPropositionsUpdate(result);
        } else if ("clearCachedPropositions".equals(call.method)) {
            Optimize.clearCachedPropositions();
            result.success(null);
        } else if ("offerDisplayed".equals(call.method)) {
            handleOfferDisplayed(call, result);
        } else if ("offerTapped".equals(call.method)) {
            handleOfferTapped(call, result);
        } else if ("generateDisplayInteractionXdm".equals(call.method)) {
            handleGenerateDisplayInteractionXdm(call, result);
        } else if ("generateTapInteractionXdm".equals(call.method)) {
            handleGenerateTapInteractionXdm(call, result);
        } else if ("generateReferenceXdm".equals(call.method)) {
            handleGenerateReferenceXdm(call, result);
        } else if ("batchDisplayed".equals(call.method)) {
            handleBatchDisplayed(call, result);
        } else if ("batchGenerateDisplayInteractionXdm".equals(call.method)) {
            handleBatchGenerateDisplayInteractionXdm(call, result);
        } else {
            result.notImplemented();
        }
    }

    @SuppressWarnings("unchecked")
    private void handleUpdatePropositions(MethodCall call, final Result result) {
        Map<String, Object> arguments = (Map<String, Object>) call.arguments;
        List<Map<String, Object>> scopesList = (List<Map<String, Object>>) arguments.get("decisionScopes");
        List<DecisionScope> scopes = FlutterAEPOptimizeDataBridge.decisionScopesFromList(scopesList);

        if (scopes == null || scopes.isEmpty()) {
            result.error("INVALID_ARGUMENT", "decisionScopes is required", null);
            return;
        }

        Map<String, Object> xdm = (Map<String, Object>) arguments.get("xdm");
        Map<String, Object> data = (Map<String, Object>) arguments.get("data");
        Double timeout = arguments.containsKey("timeout") && arguments.get("timeout") instanceof Number
                ? ((Number) arguments.get("timeout")).doubleValue() : null;

        AdobeCallbackWithError<Map<DecisionScope, OptimizeProposition>> callback =
                new AdobeCallbackWithError<Map<DecisionScope, OptimizeProposition>>() {
                    @Override
                    public void call(Map<DecisionScope, OptimizeProposition> propositions) {
                        AndroidUtil.runOnUIThread(() ->
                                result.success(FlutterAEPOptimizeDataBridge.mapFromPropositionsMap(propositions)));
                    }

                    @Override
                    public void fail(AdobeError adobeError) {
                        final AdobeError error = adobeError != null ? adobeError : AdobeError.UNEXPECTED_ERROR;
                        AndroidUtil.runOnUIThread(() ->
                                result.error(Integer.toString(error.getErrorCode()),
                                        "updatePropositions failed",
                                        error.getErrorName()));
                    }
                };

        if (timeout != null) {
            Optimize.updatePropositions(scopes, xdm, data, timeout, callback);
        } else {
            Optimize.updatePropositions(scopes, xdm, data, callback);
        }
    }

    @SuppressWarnings("unchecked")
    private void handleGetPropositions(MethodCall call, final Result result) {
        Map<String, Object> arguments = (Map<String, Object>) call.arguments;
        List<Map<String, Object>> scopesList = (List<Map<String, Object>>) arguments.get("decisionScopes");
        List<DecisionScope> scopes = FlutterAEPOptimizeDataBridge.decisionScopesFromList(scopesList);

        if (scopes == null || scopes.isEmpty()) {
            result.error("INVALID_ARGUMENT", "decisionScopes is required", null);
            return;
        }

        Double timeout = arguments.containsKey("timeout") && arguments.get("timeout") instanceof Number
                ? ((Number) arguments.get("timeout")).doubleValue() : null;

        AdobeCallbackWithError<Map<DecisionScope, OptimizeProposition>> callback =
                new AdobeCallbackWithError<Map<DecisionScope, OptimizeProposition>>() {
                    @Override
                    public void call(Map<DecisionScope, OptimizeProposition> propositions) {
                        AndroidUtil.runOnUIThread(() ->
                                result.success(FlutterAEPOptimizeDataBridge.mapFromPropositionsMap(propositions)));
                    }

                    @Override
                    public void fail(AdobeError adobeError) {
                        final AdobeError error = adobeError != null ? adobeError : AdobeError.UNEXPECTED_ERROR;
                        AndroidUtil.runOnUIThread(() ->
                                result.error(Integer.toString(error.getErrorCode()),
                                        "getPropositions failed",
                                        error.getErrorName()));
                    }
                };

        if (timeout != null) {
            Optimize.getPropositions(scopes, timeout, callback);
        } else {
            Optimize.getPropositions(scopes, callback);
        }
    }

    private void handleRegisterOnPropositionsUpdate(final Result result) {
        Optimize.onPropositionsUpdate(propositions -> {
            final Map<String, Object> encoded = FlutterAEPOptimizeDataBridge.mapFromPropositionsMap(propositions);
            AndroidUtil.runOnUIThread(() ->
                    channel.invokeMethod("onPropositionsUpdate", encoded));
        });
        result.success(null);
    }

    @SuppressWarnings("unchecked")
    private void handleOfferDisplayed(MethodCall call, Result result) {
        OptimizeProposition proposition = FlutterAEPOptimizeDataBridge.propositionFromOfferTrackingMap((Map<String, Object>) call.arguments);
        if (proposition != null && proposition.getOffers() != null && !proposition.getOffers().isEmpty()) {
            proposition.getOffers().get(0).displayed();
        }
        result.success(null);
    }

    @SuppressWarnings("unchecked")
    private void handleOfferTapped(MethodCall call, Result result) {
        OptimizeProposition proposition = FlutterAEPOptimizeDataBridge.propositionFromOfferTrackingMap((Map<String, Object>) call.arguments);
        if (proposition != null && proposition.getOffers() != null && !proposition.getOffers().isEmpty()) {
            proposition.getOffers().get(0).tapped();
        }
        result.success(null);
    }

    @SuppressWarnings("unchecked")
    private void handleGenerateDisplayInteractionXdm(MethodCall call, Result result) {
        OptimizeProposition proposition = FlutterAEPOptimizeDataBridge.propositionFromOfferTrackingMap((Map<String, Object>) call.arguments);
        if (proposition != null && proposition.getOffers() != null && !proposition.getOffers().isEmpty()) {
            result.success(proposition.getOffers().get(0).generateDisplayInteractionXdm());
        } else {
            result.success(null);
        }
    }

    @SuppressWarnings("unchecked")
    private void handleGenerateTapInteractionXdm(MethodCall call, Result result) {
        OptimizeProposition proposition = FlutterAEPOptimizeDataBridge.propositionFromOfferTrackingMap((Map<String, Object>) call.arguments);
        if (proposition != null && proposition.getOffers() != null && !proposition.getOffers().isEmpty()) {
            result.success(proposition.getOffers().get(0).generateTapInteractionXdm());
        } else {
            result.success(null);
        }
    }

    @SuppressWarnings("unchecked")
    private void handleGenerateReferenceXdm(MethodCall call, Result result) {
        OptimizeProposition proposition = FlutterAEPOptimizeDataBridge.propositionFromMap((Map<String, Object>) call.arguments);
        if (proposition != null) {
            result.success(proposition.generateReferenceXdm());
        } else {
            result.success(null);
        }
    }

    @SuppressWarnings("unchecked")
    private void handleBatchDisplayed(MethodCall call, Result result) {
        List<Map<String, Object>> items = (List<Map<String, Object>>) call.arguments;
        List<Offer> offers = new ArrayList<>();
        // Keep propositions alive — Offer uses a SoftReference to its proposition, which GC can
        // clear once the loop-scoped `prop` goes out of scope, resulting in empty tracking payloads.
        List<OptimizeProposition> propositions = new ArrayList<>();
        for (Map<String, Object> item : items) {
            OptimizeProposition prop = FlutterAEPOptimizeDataBridge.propositionFromOfferTrackingMap(item);
            if (prop != null && prop.getOffers() != null && !prop.getOffers().isEmpty()) {
                propositions.add(prop);
                offers.add(prop.getOffers().get(0));
            }
        }
        if (!offers.isEmpty()) {
            OfferUtils.displayed(offers);
        }
        result.success(null);
    }

    @SuppressWarnings("unchecked")
    private void handleBatchGenerateDisplayInteractionXdm(MethodCall call, Result result) {
        List<Map<String, Object>> items = (List<Map<String, Object>>) call.arguments;
        List<Offer> offers = new ArrayList<>();
        // Keep propositions alive — Offer uses a SoftReference to its proposition, which GC can
        // clear once the loop-scoped `prop` goes out of scope, resulting in empty tracking payloads.
        List<OptimizeProposition> propositions = new ArrayList<>();
        for (Map<String, Object> item : items) {
            OptimizeProposition prop = FlutterAEPOptimizeDataBridge.propositionFromOfferTrackingMap(item);
            if (prop != null && prop.getOffers() != null && !prop.getOffers().isEmpty()) {
                propositions.add(prop);
                offers.add(prop.getOffers().get(0));
            }
        }
        if (!offers.isEmpty()) {
            result.success(OfferUtils.generateDisplayInteractionXdm(offers));
        } else {
            result.success(null);
        }
    }
}
